[IsDeprecatedWeakRefSmartPointerException] PendingDownload should not be a smart pointer exception

geoffreygaren opened the pull request:
#### bc74bcf9e05d4dd52dc9a93dc0414acf8df0d748
<pre>
[IsDeprecatedWeakRefSmartPointerException] PendingDownload should not be a smart pointer exception
<a href="https://bugs.webkit.org/show_bug.cgi?id=279796">https://bugs.webkit.org/show_bug.cgi?id=279796</a>
<a href="https://rdar.apple.com/136113361">rdar://136113361</a>

Reviewed by Ryosuke Niwa.

* Source/WTF/wtf/RefCountedWithWeakPtr.h: Added. For now, this class is just
a convenience to do the right thing by default without extra typing. In the
future I plan to optimize by combining the two separate implementations.

* Source/WebKit/NetworkProcess/Downloads/PendingDownload.h: Make it refcounted.

Canonical link: <a href="https://commits.webkit.org/283792@main">https://commits.webkit.org/283792@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ba9b4a7be6377c587925091d5fb58b4516f86fc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67354 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46733 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19986 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71400 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18491 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54531 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18282 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54013 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12398 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70421 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42923 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58262 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/34475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39595 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15662 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16849 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/60469 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61564 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16003 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73101 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/66599 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11313 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15334 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61449 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11346 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58327 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61506 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/14931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9261 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2866 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/88368 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10233 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/42538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15572 "Found 9 new JSC stress test failures: wasm.yaml/wasm/function-tests/many-arguments-to-function.js.default-wasm, wasm.yaml/wasm/function-tests/many-arguments-to-function.js.wasm-collect-continuously, wasm.yaml/wasm/fuzz/export-function.js.default-wasm, wasm.yaml/wasm/fuzz/memory.js.wasm-bbq, wasm.yaml/wasm/stress/js-to-wasm-many-i64.js.wasm-slow-memory, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-bbq, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-eager-jettison, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-no-cjit, wasm.yaml/wasm/stress/tail-call.js.wasm-slow-memory (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43615 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44801 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43356 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WTF/wtf/RefCountedWithWeakPtr.h:
```diff
@@ -0,0 +1,40 @@
+namespace WTF {
+
+template<typename T>
+class RefCountedWithWeakPtr<T> : public CanMakeWeakPtr<T>, public RefCounted<T> {
```
Maybe we can call it `RefCountedAndCanMakeWeakPtr` to match `ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr`? Or maybe we should rename the latter to `ThreadSafeRefCountedWithWeakPtr`

geoffreygaren committed:
[IsDeprecatedWeakRefSmartPointerException] PendingDownload should not be a smart pointer exception
https://bugs.webkit.org/show_bug.cgi?id=279796
rdar://136113361

Reviewed by Ryosuke Niwa.

* Source/WTF/wtf/RefCountedWithWeakPtr.h: Added. For now, this class is just
a convenience to do the right thing by default without extra typing. In the
future I plan to optimize by combining the two separate implementations.

* Source/WebKit/NetworkProcess/Downloads/PendingDownload.h: Make it refcounted.

Canonical link: https://commits.webkit.org/283792@main
